### DISCUSSION TITLE
Run analysis template on the first deployment

### DISCUSSION THREAD

**Discussion by mirza-ali-ctct** [argoproj/argo-rollouts#2224] [General]
We are trying to run integration tests using `Job` on every deploy(including the first deploy). Looking at the docs https://argoproj.github.io/argo-rollouts/features/bluegreen/#sequence-of-events, it appears that we have to change the pod spec which makes sense for changes after first deploy. Is it possible to run `AnalysisTemplate` on first deploy?

**Comment by kostis-codefresh** [comment]
The first deployment is special because there is no "previous" deployment to rollback to. So in the first deploy Argo Rollouts does a standard deployment.

Can you explain what exactly what you want to achieve?

**Reply by davidmortiz** [reply]
We're using argo-rollouts to run tests and our use case would be the first time we deploy an application we'd like to run the analysis template to make sure the application is running properly even through there technically wouldn't be a rollout, it would give us the opportunity to know if something went wrong. 

Our use case is we want to be able to run tests when we migrate applications across clusters.

**Reply by kostis-codefresh** [reply]
I see. But then if the analysis fails what do you expect to happen? There is no previous version to rollback to. Or you expect Rollouts to keep the new version running anyway? Or maybe to shut it down completely?

**Reply by mirza-ali-ctct** [reply]
On the first deployment, we do not want to roll back as there is no previous version.

**Reply by BWagenerGenerali** [reply]
We have the exact same use-case as @davidmortiz. While it might vary case-by-case we also would like the deployment to remain if the very first AnalysisRun on the very first deployment fails.

@kostis-codefresh is this something that could be considered as a reasonable addition to the behavior of Rollouts?

**Reply by kostis-codefresh** [reply]
Feel free to open a GitHub issue about this as a feature request. 

At least to me this seems counterintuitive (running an analysis that fails and keep the deployment). Then why run the analysis in the first place if the result doesn't matter?

**Reply by kostis-codefresh** [reply]
If you are using Argo CD, you could also insert some smoke-tests in the post-sync phase.

**Reply by BWagenerGenerali** [reply]
Thank you for the quick response!

> At least to me this seems counterintuitive (running an analysis that fails and keep the deployment). Then why run the analysis in the first place if the result doesn't matter?

I understand this train of thought.

I think both davidmortiz and I come from a setup where an application can be deployed to multiple clusters, which might help explain why this is something some people ask for:

Speaking for the setup I'm working with specifically, this happens during blue-green cluster upgrades. The idea is to have a fully declaratively defined cluster including the workload that should be deployed to it. Cluster upgrades become very simple using this setup as we do not rely on in-place upgrades.

A very similar setup to ours is outlined in [this article](https://aws.amazon.com/blogs/containers/blue-green-or-canary-amazon-eks-clusters-migration-for-stateless-argocd-workloads/) in the section "Solution overview":
![R53_weighted_blue_green](https://github.com/user-attachments/assets/1e6a6330-c0ca-4fdf-a894-fef781ef2330)

After creating a new cluster and before routing production traffic to it, our devops teams can confirm everything is working as expected and then route traffic to it. This confirmation step is what's stumping our developers, as of course they expect to be able to use the same mechanism (AnalysisRuns) for smoke-tests after a blue-green cluster upgrade as they do during regular blue-green application updates.

> If you are using Argo CD, you could also insert some smoke-tests in the post-sync phase.

True - the reason for us to prefer a Rollouts-native solution is that it would only require configuring smoke-tests once in an AnalysisTemplate instead of both an AnalysisTemplate and a post-sync job. That and AnalysisTemplates are just a nicer interface for test definitions.

> Feel free to open a GitHub issue about this as a feature request.

Will do, thank you! (edit: https://github.com/argoproj/argo-rollouts/issues/3984)
